### PR TITLE
add deallocation tracking

### LINKED ISSUE #1314: Extend track alloc id to show deallocations
Currently `-Zmiri-track-alloc-id=<id>` shows a stack trace of where some memory got allocated, but not where it was deallocated. It would be nice to show where it was deallocated in order to debug use-after-free bugs.

### PR DESCRIPTION
**Pull request by KrishnaSannasi**
fixes #1314

### CODE REVIEW

**Comment by KrishnaSannasi** [comment]
cc @RalfJung

File: src/diagnostics.rs [hunk 1]
```diff
@@ -42,6 +42,7 @@ impl MachineStopType for TerminationInfo {}
 pub enum NonHaltingDiagnostic {
     PoppedTrackedPointerTag(Item),
     CreatedAlloc(AllocId),
+    DestroyAlloc(AllocId)
```
Comment (RalfJung) [inline src/diagnostics.rs:45]: Given the others use "...ed" form of the verb, I think this should "DestroyedAlloc" or maybe "FreedAlloc".

File: src/machine.rs [hunk 2]
```diff
@@ -116,7 +116,7 @@ pub struct MemoryExtra {
     pub(crate) rng: RefCell<StdRng>,
 
     /// An allocation ID to report when it is being allocated
-    /// (helps for debugging memory leaks).
+    /// (helps for debugging memory leaks and use after free bugs).
```
Comment (RalfJung) [inline src/machine.rs:119]: Please also adjust the README description of the flag.

**Comment by RalfJung** [comment]
Looks good, just some nits. :)

File: README.md [hunk 3]
```diff
@@ -209,7 +209,13 @@ Some native rustc `-Z` flags are also very relevant for Miri:
   functions.  This is needed so that Miri can execute such functions, so Miri
   sets this flag per default.
 * `-Zmir-emit-retag` controls whether `Retag` statements are emitted. Miri
-  enables this per default because it is needed for [Stacked Borrows].
+  enables this per default because it is needed for validation.
+* `-Zmiri-track-pointer-tag=<tag>` shows a backtrace when the given pointer tag
```
Comment (RalfJung) [inline README.md:213]: Something went wrong here, you duplicated the existing help texts instead of expanding it.

Comment (KrishnaSannasi) [reply README.md:213]: whoops, looks like I messed up the merge conflict

**Comment by RalfJung** [comment]
Please avoid merge commits in PRs, and rebase instead.

**Comment by KrishnaSannasi** [comment]
cc @RalfJung Ok, I've made the changes you asked and cleaned up the history. I think this is good to go.

**Comment by RalfJung** [comment]
Indeed I think so too, thank you. :)
@bors r+

**Pull request merged**